src/jinja2/tatsu.py

```diff
@@ -0,0 +1,137 @@
+import pprint
+
+from tatsu import parse
+
+from .lexer import Lexer
+from .utils import LRUCache
+
+# cache for the lexers. Exists in order to be able to have multiple
+# environments with the same lexer
+_tatsu_lexer_cache = LRUCache(50)
+
+
+def get_tatsu_lexer(environment):
+    """Return a tatsu lexer which is probably cached."""
+    key = (
+        environment.block_start_string,
+        environment.block_end_string,
+        environment.variable_start_string,
+        environment.variable_end_string,
+        environment.comment_start_string,
+        environment.comment_end_string,
+        environment.line_statement_prefix,
+        environment.line_comment_prefix,
+        environment.trim_blocks,
+        environment.lstrip_blocks,
+        environment.newline_sequence,
+        environment.keep_trailing_newline,
+    )
+    tatsu_lexer = _tatsu_lexer_cache.get(key)
+    if tatsu_lexer is None:
+        tatsu_lexer = TatsuLexer(environment)
+        _tatsu_lexer_cache[key] = tatsu_lexer
+    return tatsu_lexer
+
+
+GRAMMAR = r"""
+    @@grammar::CALC
+
+    @@whitespace:://
+
+
+    start = {expression}+ ;
+
+
+    expression
+        =
+        | code
+        | variable
+        | comment
+        | data:/(.|\n)/
+        ;
+
+    code
+        =
+        block_begin:'{%' {code_content:code_content}+ block_end:'%}'
+        ;
+
+    code_content
+        =
+        | whitespace:/[\ \n]+/
+        | name:/\w+/
+        | operator:operator
+        ;
+
+    operator
+        =
+        | '+' | '-' | '/' | '//' | '*' | '**' | '~'
+        | '[' | ']' | '(' | ')' | '{' | '}'
+        | '==' | '!=' | '>' | '>=' | '<' | '<=' | '='
+        | '.' | ':' | '|' | ',' | ';'
+        ;
+
+    variable
+        =
+        variable_begin:'{{' {variable_content:variable_content}+ variable_end:'}}'
+        ;
+
+    variable_content
+        =
+        | whitespace:/[\ \n]+/
+        | name:/\w+/
+        ;
+
+    comment
+        =
+        comment_begin:'<!--' ~ {comment_data:comment_content}+ comment_end:'-->'
+        ;
+
+    comment_content = /([^-]|-(?!->)|\\n)/ ;
+"""
+
+
+def tatsu_tokenize(source):
+    # use print for debug
+    pprint.pprint(parse(GRAMMAR, source), indent=2, width=20)
```

Comment:
Remove debug statements before submitting (okay to keep for now since it's still a WIP)